Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractRandomCutTree.java

```diff
@@ -541,28 +544,28 @@ protected boolean leftOf(double[] point, NodeReference nodeReference) {
         checkNotNull(visitorFactory, "visitor must not be null");
         checkState(rootIndex != null, "this tree doesn't contain any nodes");
         MultiVisitor<R> visitor = visitorFactory.apply(this);
-        traverseTreeMulti(point, visitor, rootIndex, nodeView, 0);
+        traverseTreeMulti(point, visitor, rootIndex, 0);
         return visitor.getResult();
     }
 
-    private <R> void traverseTreeMulti(double[] point, MultiVisitor<R> visitor, NodeReference currentNode,
-            INode<NodeReference> nodeView, int depthOfNode) {
-        if (isLeaf(currentNode)) {
-            visitor.acceptLeaf(nodeView.getNodeView(currentNode), depthOfNode);
+    private <R> void traverseTreeMulti(double[] point, MultiVisitor<R> visitor, NodeReference node, int depthOfNode) {
+        if (isLeaf(node)) {
+            visitor.acceptLeaf(getNodeView(node), depthOfNode);
         } else {
-            if (visitor.trigger(nodeView.getNodeView(currentNode))) {
-                NodeReference leftChild = getLeftChild(currentNode);
-                traverseTreeMulti(point, visitor, leftChild, nodeView.getNodeView(leftChild), depthOfNode + 1);
+            if (visitor.trigger(getNodeView(node))) {
+                traverseTreeMulti(point, visitor, getLeftChild(node), depthOfNode + 1);
                 MultiVisitor<R> newVisitor = visitor.newCopy();
```

Comment:
Is this correct? It seems weird that we make a copy of visitor after traversing the left path, and then we combine with visitor after traversing the right path. It seems like we're including the result of traversing the left path twice.

Reply:
This does allow a stronger visitor pattern instead of making a clean copy -- the pattern now also can express a preorder traversal.